blog/_posts/2021-08-20-Mitchell-Johnson_Jamie-Schmidt.md

```diff
@@ -0,0 +1,296 @@
+  
+When coding an Evaluator, the most important thing is to make sure that, for whatever inputs your evaluate function requires, you've got Brains that can supply it.  They're configured in the same way that Brains and Genomes are, by being given a list of input names that correspond to the types that the Evaluator is expecting.
+  
+We did not make many changes to the Evaluator model, aside from making it compatible with DynamicOrgs, which in turn made them significantly easier to customize since every DynamicOrg Evaluator follows the same steps: 
+  1. Get the emp::DataMap
+    - This is boilerplate for every DynamicOrg evaluator and can be copied from other evaluators. DataMap will contain all the output from the organisms brain. And the brains output is what needs to be evaluated.
+  2. Get all the input values you want from it
+    - Simply access the DataMap with GetVar
+  3. Calculate fitness from them
+    - Implement any custom function for calculating fitness. Ex: counting 1s or pattern matching
+  4. Return fitness
+    - Make sure that function actually returns your calculated fitness.
+                                      
+                                      
+And here's an example of those exact steps, for a very simple Evaluator that just counts the number of ones (or zeros) in a vector of bits.  The function in questin here is OnUpdate().
+        
+```cpp
+        
+    void OnUpdate(size_t /* update */) override {
+      emp_assert(control.GetNumPopulations() >= 1);
+
+      // Loop through the population and evaluate each organism.
+      double max_fitness = 0.0;
+      emp::Ptr<Organism> max_org = nullptr;
+      mabe::Collection alive_collect( target_collect.GetAlive() );
+      for (Organism & org : alive_collect) {        
+        // Make sure this organism has its bit sequence ready for us to access.
+        org.GenerateOutput();
+
+        emp::DataMap dm = org.GetVar<emp::DataMap>(bits_trait);
+
+        Genome::Head head = dm.Get<Genome::Head>(bits_trait);
+
+        double fitness = 0.0;
+        double total = 0.0;
+
+        while (head.IsValid()) {
+          fitness += head.ReadBit();
+          total += 1;
+        }
+
+        // If we were supposed to count zeros, subtract ones count from total number of bits.
+        if (count_type == 0) fitness = total - fitness;
+
+        // Store the count on the organism in the fitness trait.
+        org.SetVar<double>(fitness_trait, fitness);
+
+        if (fitness > max_fitness || !max_org) {
+          max_fitness = fitness;
+          max_org = &org;
+        }
+      }
+
+      std::cout << "Max " << fitness_trait << " = " << max_fitness << std::endl;
+    }
+        
+```
+        
+The important thing here (and the thing which you will likely need to edit for any custom experiment is this bit of code: 
```

Comment:
add ")"

Reply:
explain why you would need to edit/how you would edit this bit  of code

Reply:
done